### PR TITLE
Implement Eventually Consistent Mock Repository for SnapshotResiliencyTests

### PR DESCRIPTION
**Pull request by original-brownbear**
* Implement simple simulation of an eventually consistent blob repository to be able to reproduce issues resulting from #38941. See new Javadoc added here for the exact behavior simulated.
   * Added a todo to make the behavior include inconsistency over a longer period of time and specifically over multiple back to back snapshots and deletes that include node failures with concurrent indexing (currently we test this, but without changing the data in the indices snapshotted so we're not affected by the eventual consistent list operation on shard directories that would break things). 
* Fortunately, as a result of #40652 we can now merge this in as the current tests are able to deal with S3 semantics since they don't rely on consistent list operations in what they test (they don't do any updates to a shard's segment files, so it all seems to work out fine (I ran 5k iterations of each test to confirm)).

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                final boolean result = super.blobExists(blobName);
+                if (result == false) {
+                    cachedMisses.add(blobName);
+                }
+                return result;
+            }
+
+            @Override
+            public InputStream readBlob(String name) throws IOException {
+                ensureReadAfterWrite(name);
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: I'm not sure, how we're simulating eventual consistency for non-first write operation here.
If there is pendingWrite, it seems that we're just executing it and perform read of the latest pendingWrite.

Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: That's a fair point, thanks for spotting this! Practically it doesn't change anything (fortunately) since the `index.latest` blob is the only blob we ever overwrite (and we don't use that blob in normal blob repository operation) but I'll fix this shortly :)

**Comment by original-brownbear** [comment]
@andrershov fixed the incorrect consistency in  e320ebc  (I added a todo to make the simulation even less consistent, but I'm actually not sure that it's needed).

I also and (more importantly) fixed an issue with the `extra0` files that Lucene leaves around that started breaking this test now that https://github.com/elastic/elasticsearch/pull/40857 is merged. Technically that should/could be a separate PR but since it's an issue that's so much more pronounced for the async repository I put it in here since even without the `extra0`file we could run into empty trees since we have no guarantees that the delete for the directory runs after the deletes for the individual files.
This isn't a problem in a real (S3) repository because there isn't even any folders in S3 anyway.

**Comment by original-brownbear** [comment]
Lets get https://github.com/elastic/elasticsearch/pull/41332 in first before we fix the remaining things here :)

**Review by andrershov — CHANGES_REQUESTED** [review]
@original-brownbear If we truly want to emulate eventual consistency, we can just remember all writes done to some blob path as a list of byte[] in memory and each time read is executed, we can return random write.
So I would rather have  EventualConsistentInMemoryBlobContainer implementing  BlobContainer interface and use it in SnapshotResiliencyTests.
The implementation that you're proposing is hard to follow, contains bugs (as we figured out yesterday) and does not fully resemble AWS consistency model.
Also, I don't think it's a good idea to emulate only those AWS consistency phenomena that we currently support. I think we should implement proper eventual consistent blob container and plug it in, once we think all bugs (that assume stronger consistency) in our code are addressed.

**Comment by original-brownbear** [comment]
@andrershov 

> If we truly want to emulate eventual consistency, we can just remember all writes done to some blob path as a list of byte[] in memory and each time read is executed, we can return random write.

I added a todo that covers this now. I don't think there's any point in simulating this behavior in the first place. There is no mathematically sound way to handle concurrent updates to a single blob and we simply should remove all remaining places where we execute those updates. Once we fixed them, we should add the assertion that the todo asks for (see here https://github.com/elastic/elasticsearch/pull/40893/files#diff-5c02ffbdee1998c540bcfebd54c4b422R164). The only reason I'd keep `index.latest` here is that it would be a best effort thing for recovering from a full cluster loss.

> The implementation that you're proposing is hard to follow, contains bugs (as we figured out yesterday) and does not fully resemble AWS consistency model.

Simplified it a bit now + fixed the discussed bugs. See above for why simulating the full eventual consistency behavior makes no sense. We can get the same safety by asserting that we don't do actions that trigger the behavior not implemented/simulated.

> Also, I don't think it's a good idea to emulate only those AWS consistency phenomena that we currently 
support.

Well the tests implemented here cover this fix https://github.com/elastic/elasticsearch/pull/40652 for example. Not sure why we can not make incremental progress here? Especially when every step forward here is of actual value to users and every new breakage (from say a new LIST call somewhere) may have catastrophic consequences.

 I think we should implement proper eventual consistent blob container and plug it in, once we think all bugs (that assume stronger consistency) in our code are addressed.

See above. This simply isn't realistic in the short term and we should have tests covering what we can/could fix now.

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Mock Repository that simulates the eventually consistent behaviour of AWS S3 as documented in the
+ * <a href="https://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html#ConsistencyModel">AWS S3 docs</a>.
+ * Specifically this implementation simulates:
+ * <ul>
+ *     <li>First read after write is consistent for each blob. (see S3 docs for specifics)</li>
+ *     <li>Deletes and updates to a blob can become visible with a delay.</li>
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: Can you elaborate on how we're simulating it here?
From what I see, `pendingWrites` map will always contain the latest write in it and `readBlob` executes `ensureReadAfterWrite`, that will run the latest `pendingWrite` if there is no cached miss.
So delays here are emulated for cached misses only, I don't see how the read could read, for example, the first write if the second write is delayed.
Can you please give more precise description of what this implementation is doing?

**Review by andrershov — CHANGES_REQUESTED** [review]

**Comment by andrershov** [comment]
@original-brownbear in general, I'm ok to move iteratively, but in this case, we need to be explicit about: 
* What is the exact behaviour we're emulating?
* What are the phenomena we're not emulating?
* What particular adjustments should be done to the current implementation to handle these phenomena?
@ywelsch can you give your opinion on this as well?

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: > So delays here are emulated for cached misses only, I don't see how the read could read, for example, the first write if the second write is delayed.

This is not implemented and I chose to add this todo instead: https://github.com/elastic/elasticsearch/pull/40893/files#diff-5c02ffbdee1998c540bcfebd54c4b422R164

There is no point in implementing out of order updates to blobs explained above:
* Updates to blobs have no guaranteed read visibility -> our logic shouldn't rely on updates to blobs ever and we need to remove any spots where they occur and then add an assertion here that simply throws on updates.

So what we have now is simply what is listed there, the fact that two consecutive writes to a blob without a read will result in the second write result showing up consistently is addressed by the todo as well.

**Comment by original-brownbear** [comment]
> What is the exact behaviour we're emulating?

See list in the Javadoc and link to S3 docs in it.

> What are the phenomena we're not emulating?

We're not emulating out of order reads for blob updates. All other S3 eventual consistency should be emulated now.

> What particular adjustments should be done to the current implementation to handle these phenomena?

This is not clear yet. The problem is very visible in the implementation here though:

* We do not rely on any consistency guarantees beyond what S3 offers for snapshots without node failures anymore (out of order execution without advancing the deterministic task queue clock works fine). We do however require some listing consistency for master failover situations.

For the case of simple master failovers a fix for the situation is incoming in my work here https://github.com/elastic/elasticsearch/compare/master...original-brownbear:delete-lock-via-cs?expand=1 (still very rough but getting there shortly). We can simply use the cluster state as a source of consistency to track enough metadata to allow us to not have to use inconsistent list operations ever. Also, we must use the cluster state for more fine grained control over the various update steps on the repo to avoid updating individual blobs ever.

For the situation of losing the cluster state entirely and then restoring from an eventually consistent repository a completely air-tight solution does not exist yet. See last sync on the details of this, but long story short:
* There's ways of getting enough consistency from multi-part upload APIs potentially like the Hadoop mapper S3 committer does
* This is not something to be addressed in the short term since the situation can just as well be addressed by waiting a few minutes (and in practice < 1 min in 99.9% of cases) after losing the entire cluster so the gain/effort ration for implementing a likely highly complex solution for an extremely unlikely and easily preventable situation is just not there :)

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java (hunk 2 above)
Comment (andrershov) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: @original-brownbear Do these statements describe what your implementation is doing more precisely?
1) If there is no cached miss, read always returns the latest write.
2) If there is a cached miss, read might return nothing or the latest write.
3) Writes and deletes become visible to list operations with a delay, but never out of order.

Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: @andrershov sounds about the same to me :) Though I'm not so sure about the phrase "cached miss". It's a little iffy calling it that in docs when it's not a phrase used by the AWS S3 docs linked above the text.
Maybe it would make it easier if I just added a line to what is there already that simply states:

"Contrary to what is stated in the AWS S3 documentation, in this implementation writes and deletes become visible to readers in the order they were executed in. A list or get operation will never see an older state after already having seen a newer state."

?

**Comment by original-brownbear** [comment]
@ywelsch can you take a look here. I think this has quite a bit of value. E.g. https://github.com/elastic/elasticsearch/commit/f8df21787a63c29597cc9cb403db7d94e892e005 just now reproducing/fixing the eventual consistency issue explained in #41898

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

**Review by ywelsch — CHANGES_REQUESTED** [review]
We discussed this PR. @original-brownbear will explore writing an eventually consistent repository implementation that's modeling exactly S3's behavior and that's based on the history of operations that the repo has seen. This implementation can then be pluggably adapted to show that under a restricted set of conditions (i.e. a model stronger than S3's behavior) the current tests pass.

**Comment by original-brownbear** [comment]
@ywelsch I actually tried this a little and it's quite bothersome to implement without the infrastructure that comes from extending the fs repository, mainly because all the consistency checks in the resiliency tests check the files directly. With the refactoring of those that's incoming in https://github.com/elastic/elasticsearch/pull/42189 this should be doable though :)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

**Comment by original-brownbear** [comment]
@ywelsch @andrershov I've now implemented this (or at least a somewhat boiled down version of it for the time being) by doing everything in memory and tracking all previous operations.
I didn't actually implement simulating any inconsistent reads and instead just made the logic throw an `AssertionError` if they were to happen. 
This gives the same amount of coverage as the previous solution I had coded up here (on known and fixed issues that resulted from read=before-write) and can be extended to cover more situations as they are fixed. Implementing a simulation of inconsistent answers to get or exists calls really wouldn't add anything beyond what just throwing in these spots adds, it would just make it harder to understand the resulting test failure.

Technically speaking, one could nicely reproduce all the possible listing inconsistencies with this by simply randomly selecting an offset in the history of repo operations (`context.actions`) and using the blob structure at that index for the listing. But for now, there's no point in doing so because we have no code that would be able to handle an inconsistent listing. Once we have that I suppose we can turn on that kind of behavior.

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+                            writes.add(relevantAction.data);
+                        }
+                        if (writes.isEmpty() && relevantAction.operation == Operation.GET) {
+                            readBeforeWrite = true;
+                        }
+                    }
+                    if (writes.isEmpty()) {
+                        throw new NoSuchFileException(blobPath);
+                    }
+                    if (readBeforeWrite == false && writes.size() == 1) {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: this does not take deletions into account.

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+                }
+                synchronized (context.actions) {
+                    context.actions.add(new BlobStoreAction(Operation.DELETE, path.buildAsString() + blobName));
+                }
+            }
+
+            @Override
+            public void delete() {
+                if (closed.get()) {
+                    throw new AssertionError("Blobstore is closed already");
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: perhaps factor this out into an ensureNotClosed() method

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java (hunk 3 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: This shows that the class here needs unit tests that validates its behavior.

**Review by ywelsch — COMMENTED** [review]
I've asked for more tests

**Comment by original-brownbear** [comment]
@ywelsch I pushed https://github.com/elastic/elasticsearch/pull/40893/commits/d1f1e798735c6b43bffe0077d65f3ee09c8bdd13

This adds tests and fixes the behavior on deletes. Also I removed the logic giving special case handling to `incompatible-snapshots` and `index.latest` since both aren't used by the snapshot logic (except for the URL repository but that's not relevant here) so there's no point having leniency for them now.

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+                this.path = path;
+            }
+
+            @Override
+            public BlobPath path() {
+                return path;
+            }
+
+            @Override
+            public boolean blobExists(String blobName) {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: Not for this PR, but AFAICS we can get rid of this method on BlobContainer. We only use it in BlobStoreRepository to check existence of `master.dat` but we could as well just read that file, it's super small

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+                            writes.add(relevantAction.data);
+                        }
+                        if (writes.isEmpty() && relevantAction.operation == Operation.GET) {
+                            readBeforeWrite = true;
+                        }
+                    }
+                    if (writes.isEmpty()) {
+                        throw new NoSuchFileException(blobPath);
+                    }
+                    if (readBeforeWrite == false && relevantActions.size() == 1) {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: This sounds like a complicated way of just saying the first relevant action is a PUT?
Given that you're already filtering trailing GETs from relevantActions. Why not just check here if `relevantActions.size() == 1 && relevantActions.get(0).operation == PUT`?

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+                    boolean readBeforeWrite = false;
+                    for (BlobStoreAction relevantAction : relevantActions) {
+                        if (relevantAction.operation == Operation.PUT) {
+                            writes.add(relevantAction.data);
+                        }
+                        if (writes.isEmpty() && relevantAction.operation == Operation.GET) {
+                            readBeforeWrite = true;
+                        }
+                    }
+                    if (writes.isEmpty()) {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: maybe simpler: `relevantAction.stream().noneMatch(a -> a.operation == Operation.PUT)`

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+                    }
+                    throw new AssertionError("Inconsistent read on [" + blobPath + ']');
+                }
+            }
+
+            @Override
+            public void deleteBlob(String blobName) {
+                ensureNotClosed();
+                synchronized (context.actions) {
+                    context.actions.add(new BlobStoreAction(Operation.DELETE, path.buildAsString() + blobName));
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: `path.add(blobName).buildAsString()`?

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+                return Maps.ofEntries(
+                    listBlobs().entrySet().stream().filter(entry -> entry.getKey().startsWith(blobNamePrefix)).collect(Collectors.toList())
+                );
+            }
+
+            @Override
+            public void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists)
+                throws IOException {
+                ensureNotClosed();
+                // TODO: Throw if we try to overwrite any blob other than incompatible_snapshots or index.latest with different content
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: what about this TODO? Can we enable this already for some files?

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepositoryTests.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            blobStoreContext.forceConsistent();
+            assertMissing(blobContainer, blobName);
+        }
+    }
+
+    private static void assertThrowsOnInconsistentRead(BlobContainer blobContainer, String blobName) throws IOException {
+        try (InputStream in = blobContainer.readBlob(blobName)) {
+            fail("Inconsistent read should throw");
+        } catch (AssertionError assertionError) {
+            assertThat(assertionError.getMessage(), equalTo("Inconsistent read on [" + blobName + ']'));
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepositoryTests.java]: use expectThrows?

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepositoryTests.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        }
+    }
+
+    private static void assertMissing(BlobContainer container, String blobName) throws IOException {
+        try (InputStream in = container.readBlob(blobName)) {
+            fail("Reading a non-existent blob should throw");
+        } catch (NoSuchFileException expected) {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepositoryTests.java]: expectThrows

**Review by ywelsch — COMMENTED** [review]
I've left some more comments.

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java (hunk 9 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: Yea we can for all but `snap-` (at the root level) and `index-N` at the root level I think. Will do :)

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: Yea you're right. The main reason I kept it was actually that Azure used it as well for some magic with the secondary dns failover but obviously that can become an implementation detail there :) I put it on my list.

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepositoryTests.java (hunk 11 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepositoryTests.java]: Done in 58a9d4234b6fe151743bb3bbf842b5ef704e6ba9

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepositoryTests.java (hunk 10 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepositoryTests.java]: Done in 58a9d4234b6fe151743bb3bbf842b5ef704e6ba9

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java (hunk 9 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: I way beefed this up now. We were only running into this for `snap-` with the current implementation (since we don't have list inconsistencies simulated yet index-N isn't an issue). 
For the `snap-` I actually made it check for the case when whatever is written only differs in timstamps (on a master failover this can happen) and let it pass.
Shard level overwrites I made fail 100% of the time (see comment I added for that, it's not expected).
I think this is pretty neat now. This hasn't failed over 10k+ runs of `SnapshotResiliencyTests` and shows that we're handling master fail-overs reasonably safely on S3 (obviously not testing list inconsistencies yet ... but still :D).

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java (hunk 7 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: Done :)

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java (hunk 6 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: True, sorry for the noise. Simplified this now.

**Comment by original-brownbear** [comment]
@ywelsch thanks for taking a look. I fixed the dirty parts you pointed out and way enhanced the overwrite check + added test coverage for that now. Let me know what you think.

File: server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java]: https://github.com/elastic/elasticsearch/pull/44472

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
thanks @ywelsch !
